### PR TITLE
IBP-4338-BrapiGetProgramsV2

### PR DESCRIPTION
**Pull request by gelli-leafnode**
New endpoint for Brapi get programs v2 
Retain current behavior programDbId = program_uuid 
Implement ProgramSearchRequest 
Created new fields under Programs pojo for brapiv2 
Create unit test and edit existing get programs v1

### CODE REVIEW

File: src/main/java/org/ibp/api/brapi/v1/program/Program.java [hunk 1]
```diff
@@ -2,21 +2,41 @@
...
 	private String abbreviation;
 	private String objective;
+
+	@JsonView({BrapiView.BrapiV1_2.class, BrapiView.BrapiV1_3.class})
+	private String name;
+	@JsonView({BrapiView.BrapiV1_2.class, BrapiView.BrapiV1_3.class})
 	private String leadPerson;
 
+	@JsonView(BrapiView.BrapiV2.class)
+	private String programName;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/brapi/v1/program/Program.java:26]: provide a new line after each property

File: src/main/java/org/ibp/api/brapi/v1/program/ProgramMapper.java [hunk 2]
```diff
@@ -8,6 +9,7 @@
 public class ProgramMapper {
 
 	private static ModelMapper applicationWideModelMapper = ApiMapper.getInstance();
+	private static JacksonAnnotationIntrospector annotationIntrospector;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/brapi/v1/program/ProgramMapper.java:12]: What is this for?

File: src/main/java/org/ibp/api/brapi/v2/program/ProgramResourceBrapi.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    @Autowired
+    private ProgramService programService;
+
+    @ApiOperation(value = "Get filtered list of breeding Programs", notes = "Get a filtered list of breeding Programs. This list can be filtered by common crop name to narrow results to a specific crop.")
+    @RequestMapping(value = "/{commonCropName}/brapi/v2/programs", method = RequestMethod.GET)
+    @ResponseBody
+    @JsonView(BrapiView.BrapiV2.class)
+    public ResponseEntity<EntityListResponse<ProgramDetailsDto>> listPrograms(
+            @ApiParam(value = "Filter by the common crop name. Exact match.")
+            @PathVariable final String commonCropName,
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/brapi/v2/program/ProgramResourceBrapi.java:43]: As discussed,
1.  commonCropName is a filter not URL path variable
2. retrieve list of crops accessible to logged in user and only return programs for those crop/s. Perform validation that commonCropName paramter, if existing, is part of that list of crops

File: src/main/java/org/ibp/api/java/program/ProgramService.java [hunk 4]
```diff
@@ -15,9 +14,9 @@
 
 	List<ProgramSummary> listProgramsByCropNameAndUser(WorkbenchUser user, String cropName);
 
-	List<ProgramDetailsDto> getProgramsByFilter(final int pageNumber, final int pageSize, final Map<ProgramFilters, Object> filters);
+	List<ProgramDetailsDto> getProgramsByFilter(final int pageNumber, final int pageSize, final ProgramSearchRequest programSearchRequest);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/program/ProgramService.java:17]: Use Pageable interface

File: src/main/java/org/ibp/api/brapi/v1/program/Program.java [hunk 5]
```diff
@@ -2,21 +2,41 @@
...
+	@JsonView(BrapiView.BrapiV2.class)
+	private String leadPersonName;
+	@JsonView(BrapiView.BrapiV2.class)
+	private String documentationURL;
+	@JsonView(BrapiView.BrapiV2.class)
+	private String externalReferences;
+	@JsonView(BrapiView.BrapiV2.class)
+	private String commonCropName;
+	@JsonView(BrapiView.BrapiV2.class)
+	private ProgramAdditionalInfo additionalInfo;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/brapi/v1/program/Program.java:38]: I don't think you need ProgramAdditionalInfo class. can be just Map<String, String>

**Review by abatac — CHANGES_REQUESTED** [review]

**Pull request merged**